### PR TITLE
Fix tied parameters

### LINKED ISSUE #6139: Integrate NeMo Automodel Qwen3.5-35B-A3B recipe with MFSDP v2
## Summary

The NeMo Automodel Qwen3.5-35B-A3B VLM fine-tuning recipe currently uses PyTorch FSDP2 and needs a validated Megatron Core path using experimental MFSDP v2.

Reference recipe:
- https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/vlm_finetune/qwen3_5_moe/qwen3_5_35b.yaml

## Reproduction / starting point

The reference recipe runs with:

```bash
automodel examples/vlm_finetune/qwen3_5_moe/qwen3_5_35b.yaml --nproc-per-node 8
```

Use the same model and the closest equivalent data, parallelism, and checkpoint settings through the MFSDP v2 path.

### PR DESCRIPTION
**Pull request by wujingyue**
## What changed

- collect MFSDP v2 parameter registrations without deduplicating names, while representing each physical `Parameter` once
- retain every owning-module-relative FQN and use the complete tuple when swapping sharded and unsharded parameter replacements
- register one gradient hook per physical trainable parameter and track backward readiness with a resettable counter
- add a tied-embedding regression test covering two forward/backward cycles, balanced annotations, DBuffer size, and optimizer-visible parameter count

## Why

PyTorch's default parameter enumeration deduplicates tied parameters by object identity. MFSDP v2 instead collected registrations by FQN, so the same physical parameter could receive multiple buffer entries and replacement objects. That broke weight identity and prevented the ready-parameter set from reaching the duplicated trainable-parameter count during backward.

## Impact

MFSDP v2 now preserves tied parameter aliases while allocating, sharding, reducing, and exposing only one physical optimizer parameter.

## Validation

- `ruff check` on all changed Python files
- Black check on all changed Python files
- Python bytecode compilation on all changed Python files
- `python -m torch.distributed.run --nproc-per-node 2 -m pytest -q tests/unit_tests/distributed/mfsdp_v2`
  - 53 passed
  - 11 skipped because they require 4-5 ranks

## Related issue

Related to #6139.

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py [hunk 1]
```diff
@@ -361,7 +360,7 @@ def _collect_owned_parameters(root_module: nn.Module) -> dict[str, nn.Parameter]
     parameters: dict[str, nn.Parameter] = {}
 
     def visit(submodule: nn.Module, submodule_fqn: str) -> None:
-        direct_parameters = list(submodule.named_parameters(recurse=False))
+        direct_parameters = submodule.named_parameters(recurse=False, remove_duplicate=False)
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:363]: It looks like tied parameters may be traversed multiple times. Would we create duplicate copies for those tied parameters? 🤔

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:363]: Yes, this is intentional; otherwise, set_module_parameter will miss some references.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py [hunk 2]
```diff
@@ -85,7 +85,7 @@ class FsdpModule:
     _name: str | None
     _parameter_groups: tuple[FsdpParameterGroup, ...]
     _context: FsdpContext | None
-    _ready_grad_parameters: set[nn.Parameter]
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:88]: I forgot why I used a set. A counter seems sufficient.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py (hunk 1 above)
Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:363]: Yeah, I feel like this would break the semantics of the tie parameter, changing it from a single parameter to multiple parameters that are updated separately. I understand that the tie parameter needs to accumulate gradients multiple times in one backward pass and perform grad reduce after the final accumulation.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:363]: I guess you are missing the fact that `remove_duplicate=False` returns a dict from potentially **multiple** strs to **the same Parameter object**. 

> changing it from a single parameter to multiple parameters that are updated separately

No, the code doesn't do that. 

> I understand that the tie parameter needs to accumulate gradients multiple times in one backward pass and perform grad reduce after the final accumulation.

Yes, the code does exactly that.

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:363]: I see, but I think there's still a correctness issue here. When constructing the tied shared parameters, they should share the same underlying weight — however, looking at [parameter_group.py L166](https://github.com/NVIDIA/Megatron-LM/pull/6164/changes#diff-318602037e8dca86bd2f6aefe51cb128938ac95d1f01feff501dc9e83520a3b4L166), the sharded parameters for tied shared parameters are each derived from `main_weight.get_dtensor(index)`, which means they could end up backed by independent main_weight entries rather than sharing the same one.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:363]: I guess you missed 
```
        for index, (parameter, fqns) in enumerate(parameter_to_fqns.items()):
```
from https://github.com/NVIDIA/Megatron-LM/pull/6164/changes#diff-318602037e8dca86bd2f6aefe51cb128938ac95d1f01feff501dc9e83520a3b4R171

This loop never calls `get_dtensor(index)` twice for the same index. cc @shjwudp

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:363]: Sorry for the miss. This mapping does work—looks great to me!

**Review by shjwudp — APPROVED** [review]
LGTM

**Review by jaredcasper — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]
